### ISSUE TITLE
GRPO: dapo/cispo/vespo loss denominator ignores `mask_truncated_completions` (the numerator doesn't)

### ISSUE THREAD

**Issue by SushantDaga** [huggingface/trl#6369]
### Reproduction

`mask_truncated_completions=True` is supposed to fully exclude truncated completions (ones that ran out of room before hitting EOS) from the loss: the docstring says "truncated completions are excluded from the loss calculation.". The reporting for it was fixed in about a year old issue+PR, but calculations never were ( #3476  and #3482 )

For `loss_type` in `["dapo", "cispo", "vespo"]`, the loss is `sum(per_token_loss) / num_items_in_batch`. Right now, only the **numerator** actually excludes truncated tokens. The **denominator** (`num_items_in_batch`) does not, it's computed earlier in `_generate()`, before `mask_truncated_completions` zeroes out those tokens' mask. So a truncated completion's tokens get counted in the denominator (diluting the average) while contributing nothing to the numerator.
 
Issue was discovered while working on the issue (#6367 ) and PR (#6368 ).
### Concrete example

Say a batch has 30 completion tokens total, and 5 of them belong to a truncated completion that `mask_truncated_completions` is supposed to drop entirely.

- Numerator: only sums over the 25 non-truncated tokens (correct).
- Denominator: still 30 (the pre-masking count) instead of 25.

The loss ends up `sum / 30` instead of `sum / 25` at about 17% smaller than it should be, and this shrinks further as more of a batch gets truncated. It's a silent scale error, not a crash, so it's easy to miss.

### Reproduction (real trainer, `loss_type="dapo"`)

With `mask_truncated_completions=True` and a short `max_completion_length` (so the model reliably produces truncated, non-EOS-terminated completions), spying on `_compute_loss`:

```
completions/clipped_ratio: 1        (every row truncated)
num_items_in_batch (denominator):   24
true post-truncation-mask count:    0
per-row completion_mask sums:       [0, 0, 0, 0, 0, 0]
```

The denominator the code actually divides by (24) is nonzero, while the true active-token count after truncation-masking is 0.

A hand-built mixed batch (half the rows truncated) quantifies the exact effect:

```
num_items_in_batch used by the real code path: 24.0
honest post-truncation-masking token count:    12.0
loss with the code's actual denominator:  -0.5
loss with an honest denominator:          -1.0
ratio: 0.5, exactly matching honest_count / raw_count (12/24)
```

The loss magnitude is silently scaled by `honest_token_count / raw_token_count` i.e. truncated tokens are excluded from the sum but still counted in "how many tokens we're averaging over."

### Why I think this is a real, unintended bug (not "working as designed")

#3476 hit the exact same root cause (computing something from `completion_mask` before `mask_truncated_completions` zeroes it) for the **logging** metrics, and got fixed there, the fix explicitly kept a copy of the mask before truncation-masking so logging stays accurate. That fix never touched the **loss denominator** path, which has the identical timing problem.

(Not the same issue as #5619 / PR #6024, which is about `num_items_in_batch` not matching the gradient-accumulation window when `steps_per_generation != gradient_accumulation_steps` a different root cause but same variable. Worth being aware of both if touching this code.)

### Where the code is

- Numerator uses the truncation-masked `completion_mask` (correctly): `mask = completion_mask if "tool_mask" not in inputs else completion_mask * inputs["tool_mask"]` in `_compute_loss`.
- Denominator (`num_items_in_batch`) is computed from `completion_lengths`/`agg_completion_lengths` inside `_generate()`, **before** `mask_truncated_completions` zeroes `completion_mask` for truncated rows (that zeroing happens later, in `_generate_and_score_completions`).

Checked for compensation elsewhere in the pipeline: none found. `GRPOTrainer.compute_loss` accepts the `num_items_in_batch` argument the HF `Trainer` base loop computes automatically, but ignores it and then reads its own internal, pre-truncation-masking value from `inputs["num_items_in_batch"]` instead.

### Suggested fix

Not as simple as "just compute `num_items_in_batch` after masking" since that variable is dual-use: the same value also feeds `state.num_input_tokens_seen` (the training-throughput counter), which *should* reflect tokens actually generated, truncated or not. Changing `num_items_in_batch` itself would silently start undercounting throughput too.

The shape of the fix that avoids that: keep a separate, masking-aware token count specifically for the `dapo`/`cispo`/`vespo` loss denominator, distinct from the one used for `num_input_tokens_seen` which mirrors how #3476 kept a separate copy of the mask for logging rather than changing the shared one. Happy to send a patch either way, just didn't want to hand you a fix that trades one silent-divergence bug for another.

### Scope note

This affects `dapo`, `cispo`, and `vespo` whenever `mask_truncated_completions=True` and at least one completion in the batch gets truncated.

### AI disclosure
same as attached PR (#6368 )

### System Info

- Platform: macOS-26.5.1-arm64-arm-64bit
- Python version: 3.12.6
- TRL version: 1.9.0.dev0+69f5907
- PyTorch version: 2.12.0
- accelerator(s): MPS
- Transformers version: 5.11.0
- Accelerate version: 1.14.0
- Accelerate config: not found
- Datasets version: 5.0.0
- HF Hub version: 1.19.0
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Liger-Kernel version: not installed
- PEFT version: 0.19.1
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by Functionhx** [comment]
I'd like to work on this. I'll submit a fix with regression tests.
